### PR TITLE
Feature: add bundled services samples from jinglundong

### PR DESCRIPTION
**Pull request by engelke**
## Description

Replaces PR #7891

The PR needs changes so created this new one within the python-docs-samples repo for ease of work.

### CODE REVIEW

**Review by m-strzelczyk — APPROVED** [review]

File: appengine/standard_python3/bundled-services/blobstore/README.md [hunk 1]
```diff
@@ -0,0 +1,35 @@
+# Access bundled Blobstore services API in Python 3
+
+This folder contains functionally identical apps that
```
Comment (ace-n) [inline appengine/standard_python3/bundled-services/blobstore/README.md:3]: Incomplete sentence?

File: appengine/standard_python3/bundled-services/blobstore/flask/main_test.py [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+    project and version number so tests can invoke it, then delete it.
+    """
+
+    output = subprocess.run(
+        "gcloud app deploy --no-promote --quiet --format=json",
+        capture_output=True,
+        shell=True,
+    )
+
+    result = json.loads(output.stdout)
```
Comment (ace-n) [inline appengine/standard_python3/bundled-services/blobstore/flask/main_test.py:35]: Ditto - print out `output.stderr`?

File: appengine/standard_python3/bundled-services/blobstore/flask/main_test.py [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+    version_id = result["versions"][0]["id"]
+    project_id = result["versions"][0]["project"]
```
Comment (ace-n) [inline appengine/standard_python3/bundled-services/blobstore/flask/main_test.py:37]: What if `result["versions"]` is empty? (Should we print out `result.stdout` for debugging in that case?)

File: appengine/standard_python3/bundled-services/blobstore/wsgi/main.py [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+    def __init__(self, routes):
+        self.routes = routes
+
+    def not_found(self, environ, start_response):
+        start_response("404 Not Found", [("Content-Type", "text/plain")])
+        return ["404 Not Found"]
+
+    def __call__(self, environ, start_response):
+        path = environ.get("PATH_INFO", "").lstrip("/")
+        print(environ)
```
Comment (ace-n) [inline appengine/standard_python3/bundled-services/blobstore/wsgi/main.py:95]: I assume this is intended (and not just leftover debugging code).

If so, maybe we should add a prefix? e.g:
```suggestion
        print("PATH_INFO value:", environ)
```

File: appengine/standard_python3/bundled-services/blobstore/wsgi/main.py [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+        for regex, handler in self.routes:
+            match = re.search(regex, path)
+            if match is not None:
+                environ["app.url_args"] = match.groups()
+                print("match.groups for %s: " % match, match.groups())
```
Comment (ace-n) [inline appengine/standard_python3/bundled-services/blobstore/wsgi/main.py:100]: Ditto - this is **not** debugging code, correct?

File: appengine/standard_python3/bundled-services/deferred/README.md [hunk 6]
```diff
@@ -0,0 +1,35 @@
+# Access bundled Deferred services API in Python 3
+
+This folder contains functionally identical apps that
```
Comment (ace-n) [inline appengine/standard_python3/bundled-services/deferred/README.md:3]: that do what?

File: appengine/standard_python3/bundled-services/deferred/wsgi/main.py [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+}
+
+
+class WSGIApplication:
+    def __call__(self, environ, start_response):
+        path = environ.get("PATH_INFO", "").lstrip("/")
+        for regex, handler in routes.items():
+            match = re.search(regex, path)
+            if match is not None:
+                return handler(environ, start_response)
```
Comment (ace-n) [inline appengine/standard_python3/bundled-services/deferred/wsgi/main.py:76]: Should this method return anything (or throw an exception) if `match` is always `None`?

(Not sure how relevant that case is.)

File: appengine/standard_python3/bundled-services/deferred/wsgi/main_test.py [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+    """
+
+    output = subprocess.run(
+        "gcloud app deploy --no-promote --quiet --format=json",
+        capture_output=True,
+        shell=True,
+    )
+
+    result = json.loads(output.stdout)
+    version_id = result["versions"][0]["id"]
```
Comment (ace-n) [inline appengine/standard_python3/bundled-services/deferred/wsgi/main_test.py:36]: Print out `output.stdout` and `output.stderr` for debuggability.

File: appengine/standard_python3/bundled-services/deferred/wsgi/main_test.py [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+        shell=True,
+    )
+
+
+def test_upload_and_view(version):
+    project_id, version_id = version
+    version_hostname = f"{version_id}-dot-{project_id}.appspot.com"
+
+    # Initial value of counter should be 0
+    response = requests.get(f"https://{version_hostname}/counter/get")
```
Comment (ace-n) [inline appengine/standard_python3/bundled-services/deferred/wsgi/main_test.py:53]: Nit: should we include the `https://` prefix in the `version_hostname` variable?

File: appengine/standard_python3/bundled-services/mail/django/main_test.py [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+            "email": f"valid-user@{version_id}-dot-{project_id}.appspotmail.com",
+            "body": "This message should be delivered",
+        },
+    )
+
+    assert response.status_code == 201
+    assert "Successfully sent mail" in response.text
+
+    # Give the mail some time to be delivered and logs to post
+    time.sleep(30)
```
Comment (ace-n) [inline appengine/standard_python3/bundled-services/mail/django/main_test.py:70]: Is it appropriate (or overkill) to support retries/exponential backoff here?

(I think some of our log-based tests do that elsewhere, to reduce flakiness.)

File: appengine/standard_python3/bundled-services/mail/flask/main_test.py [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+    """
+
+    output = subprocess.run(
+        "gcloud app deploy --no-promote --quiet --format=json",
+        capture_output=True,
+        shell=True,
+    )
+
+    result = json.loads(output.stdout)
+    version_id = result["versions"][0]["id"]
```
Comment (ace-n) [inline appengine/standard_python3/bundled-services/mail/flask/main_test.py:36]: Ditto - print `output.stdout` and `output.stderr` for debuggability purposes.

File: appengine/standard_python3/bundled-services/mail/wsgi/main.py [hunk 12]
```diff
@@ -0,0 +1,139 @@
...
+    # Return suitable response
+    response = http.HTTPStatus.OK
+    start_response(f"{response.value} {response.phrase}", [])
+    return ["success".encode("utf-8")]
+
+
+def InvalidMailSender(environ, start_response):
+    # Send invalid mail to trigger a bounce notification.
+    mail.send_mail(
+        sender="test-python-user@shreejad-knative-dev.appspotmail.com",
```
Comment (ace-n) [inline appengine/standard_python3/bundled-services/mail/wsgi/main.py:60]: Do we want the `shreejad-knative-dev` bit in here, or should we use a more generic "placeholder" value?

File: appengine/standard_python3/bundled-services/blobstore/wsgi/main.py (hunk 4 above)
Comment (engelke) [reply appengine/standard_python3/bundled-services/blobstore/wsgi/main.py:95]: I suppose so. I didn't edit this part of the code. I'll clean it up.

File: appengine/standard_python3/bundled-services/blobstore/README.md (hunk 1 above)
Comment (engelke) [reply appengine/standard_python3/bundled-services/blobstore/README.md:3]: Fixed

File: appengine/standard_python3/bundled-services/blobstore/django/main_test.py [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+    project and version number so tests can invoke it, then delete it.
+    """
+
+    output = subprocess.run(
+        "gcloud app deploy --no-promote --quiet --format=json",
+        capture_output=True,
+        shell=True,
+    )
+
+    result = json.loads(output.stdout)
```
[the remark this answers is not in the document]
Comment (engelke) [reply appengine/standard_python3/bundled-services/blobstore/django/main_test.py:35]: Done

File: appengine/standard_python3/bundled-services/blobstore/flask/main_test.py (hunk 2 above)
Comment (engelke) [reply appengine/standard_python3/bundled-services/blobstore/flask/main_test.py:35]: Done

Comment (engelke) [reply appengine/standard_python3/bundled-services/blobstore/flask/main_test.py:35]: Done

File: appengine/standard_python3/bundled-services/blobstore/flask/main_test.py (hunk 3 above)
Comment (engelke) [reply appengine/standard_python3/bundled-services/blobstore/flask/main_test.py:37]: Done

File: appengine/standard_python3/bundled-services/blobstore/wsgi/main.py (hunk 4 above)
Comment (engelke) [reply appengine/standard_python3/bundled-services/blobstore/wsgi/main.py:95]: Fixed

File: appengine/standard_python3/bundled-services/blobstore/wsgi/main.py (hunk 5 above)
Comment (engelke) [reply appengine/standard_python3/bundled-services/blobstore/wsgi/main.py:100]: It seems to be. Removed.

File: appengine/standard_python3/bundled-services/deferred/flask/main_test.py [hunk 14]
```diff
@@ -0,0 +1,83 @@
...
+    """
+
+    output = subprocess.run(
+        "gcloud app deploy --no-promote --quiet --format=json",
+        capture_output=True,
+        shell=True,
+    )
+
+    result = json.loads(output.stdout)
+    version_id = result["versions"][0]["id"]
```
[the remark this answers is not in the document]
Comment (engelke) [reply appengine/standard_python3/bundled-services/deferred/flask/main_test.py:36]: Done

File: appengine/standard_python3/bundled-services/deferred/README.md (hunk 6 above)
Comment (engelke) [reply appengine/standard_python3/bundled-services/deferred/README.md:3]: fixed

File: appengine/standard_python3/bundled-services/deferred/wsgi/main.py (hunk 7 above)
Comment (engelke) [reply appengine/standard_python3/bundled-services/deferred/wsgi/main.py:76]: That would be a 404 Not Found. Fixed.

File: appengine/standard_python3/bundled-services/deferred/wsgi/main_test.py (hunk 8 above)
Comment (engelke) [reply appengine/standard_python3/bundled-services/deferred/wsgi/main_test.py:36]: Done

File: appengine/standard_python3/bundled-services/deferred/wsgi/main_test.py (hunk 9 above)
Comment (engelke) [reply appengine/standard_python3/bundled-services/deferred/wsgi/main_test.py:53]: No, `https` is the scheme, not part of the host name in the URL.

File: appengine/standard_python3/bundled-services/mail/django/main_test.py (hunk 10 above)
Comment (engelke) [reply appengine/standard_python3/bundled-services/mail/django/main_test.py:70]: I think that would be overkill. 30 seconds is pretty long already.

File: appengine/standard_python3/bundled-services/mail/flask/main_test.py [hunk 15]
```diff
@@ -0,0 +1,88 @@
...
+    )
+
+    assert response.status_code == 201
+    assert "Successfully sent mail" in response.text
+
+    # Give the mail some time to be delivered and logs to post
+    time.sleep(30)
+
+    # Fetch logs to check messages on received mail
+    output = subprocess.run(
```
[the remark this answers is not in the document]
Comment (engelke) [reply appengine/standard_python3/bundled-services/mail/flask/main_test.py:73]: Same answer

File: appengine/standard_python3/bundled-services/mail/wsgi/main.py (hunk 12 above)
Comment (engelke) [reply appengine/standard_python3/bundled-services/mail/wsgi/main.py:60]: That should not be there! I thought I'd fixed that. I've removed it and the route to it. To send a mail you want bounced, use the same method as sending a valid mail, just to a bad address. Fixed now.

File: appengine/standard_python3/bundled-services/mail/wsgi/main.py [hunk 16]
```diff
@@ -0,0 +1,139 @@
...
+        start_response(f"{response.value} {response.phrase}", [])
+        return [html.encode("utf-8")]
+
+    else:  # POST request
+        project_id = os.environ.get("GOOGLE_CLOUD_PROJECT")
+
+        form = cgi.parse(environ["wsgi.input"])
+
+        address = form.get("email")[0]
+        if address is None:
```
[the remark this answers is not in the document]
Comment (engelke) [reply appengine/standard_python3/bundled-services/mail/wsgi/main.py:104]: Yes, if somebody uses curl to send a POST request, for example. But that will result in a 500 error. No one using this over a web browser would encounter that, so I'm not worrying about it for an example of mail.

File: appengine/standard_python3/bundled-services/mail/wsgi/main.py [hunk 17]
```diff
@@ -0,0 +1,139 @@
...
+    "": HomePage,
+}
+
+
+class WSGIApplication:
+    def __call__(self, environ, start_response):
+        path = environ.get("PATH_INFO", "")
+        for regex, callable in routes.items():
+            match = re.search(regex, path)
+            if match is not None:
```
[the remark this answers is not in the document]
Comment (engelke) [reply appengine/standard_python3/bundled-services/mail/wsgi/main.py:135]: Added a 404 Not Found default response.

File: appengine/standard_python3/bundled-services/deferred/flask/main.py [hunk 18]
```diff
@@ -0,0 +1,62 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import os
+
+from flask import Flask, request
+from google.appengine.api import wrap_wsgi_app
```
[the remark this answers is not in the document]
Comment (engelke) [reply appengine/standard_python3/bundled-services/deferred/flask/main.py:18]: The linter is happy with that's there. I'm not clear on what alternative you're proposing.

File: appengine/standard_python3/bundled-services/deferred/flask/main_test.py [hunk 19]
```diff
@@ -0,0 +1,83 @@
...
+    output = subprocess.run(
+        f"gcloud app versions delete {version_id}",
+        capture_output=True,
+        shell=True,
+    )
+
+
+def test_upload_and_view(version):
+    project_id, version_id = version
+    version_hostname = f"{version_id}-dot-{project_id}.appspot.com"
```
[the remark this answers is not in the document]
Comment (engelke) [reply appengine/standard_python3/bundled-services/deferred/flask/main_test.py:50]: Same answer. That's the scheme, not the host name.

File: appengine/standard_python3/bundled-services/mail/wsgi/main_test.py [hunk 20]
```diff
@@ -0,0 +1,94 @@
...
+    output = subprocess.run(
+        f"gcloud app versions delete {version_id}",
+        capture_output=True,
+        shell=True,
+    )
+
+
+def test_send_receive(version):
+    project_id, version_id = version
+    version_hostname = f"{version_id}-dot-{project_id}.appspot.com"
```
[the remark this answers is not in the document]
Comment (engelke) [reply appengine/standard_python3/bundled-services/mail/wsgi/main_test.py:50]: Same answer

File: appengine/standard_python3/bundled-services/blobstore/django/main_test.py [hunk 21]
```diff
@@ -0,0 +1,70 @@
...
+    )
+
+    try:
+        result = json.loads(output.stdout)
+        version_id = result["versions"][0]["id"]
+        project_id = result["versions"][0]["project"]
+    except Exception as e:
+        print(f"New version deployment output not usable: {e}")
+        print(f"Command stderr is '{output.stderr}'")
+        raise ValueError
```
Comment (ace-n) [inline appengine/standard_python3/bundled-services/blobstore/django/main_test.py:42]: Can we pass `output.stderr` to this?

(It doesn't look like that gets captured in our CI systems, e.g. [here](https://source.cloud.google.com/results/invocations/9fa7329f-3926-4ced-81fd-2689e411ce62/targets/github%2Fpython-docs-samples%2Fappengine%2Fstandard_python3%2Fbundled-services%2Fblobstore%2Fdjango/tests).)

**Pull request merged**